### PR TITLE
Derivative.d now handles negative axis arguments

### PR DESCRIPTION
**Pull request by yb6599**
Hello, I am differentitating multidimensional arrays, I found that `Derivative.d` does not handle negative axis arguments. This PR introduces a small change that makes it handle negative arguments.

For example, using PySINDy SINDYDerivative,
```
t = np.arange(3)
x = np.ones((2, 3, 2))
differentiator = ps.SINDyDerivative(kind='finite_difference', k=1, axis=-2)
d = differentiator._differentiate(x, t)
```
I run into this error that originates at these lines in `derivative.differentiation.py`:
```
[251] extra_dims = tuple(length for ax, length in enumerate(orig_shape) if ax != axis)
[252] moved_shape = (orig_shape[axis],) + extra_dims
[253] dX = np.moveaxis(dX.T.reshape((moved_shape)), 0, axis)
[254] return dX

ValueError: cannot reshape array of size 12 into shape (3,2,3,2)
```
However, after the change introduced in this PR, it won't be an issue.

### CODE REVIEW

**Comment by andgoldschmidt** [comment]
Good catch, we definitely want this ability. Would you be willing to add a small test with this PR in case of future changes? It can be added to `tests/test_interface.py`.

**Comment by andgoldschmidt** [comment]
Also, it looks like the checks failed when trying to build docs and run tests, but I bet that's on our end.

**Comment by yb6599** [comment]
@andgoldschmidt thanks for your reply, I've added the test.

**Comment by andgoldschmidt** [comment]
The changes look good, thanks! I am not sure why the github actions are failing here, but as soon as I debug that, I'll merge.

**Comment by andgoldschmidt** [comment]
@Jacob-Stevens-Haas these actions fail with `ModuleNotFoundError: No module named 'importlib_metadata'`. In [this run](https://github.com/andgoldschmidt/derivative/actions/runs/9489883664/job/26152126351), I added  `importlib_metadata` directly instead of `importlib.metadata`, but this resulted in the screenshot errors (which look like missing features due to a version or function mismatch). Is there any obvious reason why the original action is failing? 

Screenshot 2024-06-12 at 4 24 12 PM

On top of these issues, I was inspecting locally, and it looked like pytest didn't populate the entry points, so that this test failed:
```Python 
def test_hyperparam_entrypoint(): ...
```

**Comment by Jacob-Stevens-Haas** [comment]
Ok interesting.  Broadly, `importlib.metadata` contains new metadata features that are provided to older python releases via the pip package `importlib_metadata`.  I wanted to use a version of `importlib.metadata` that came out in 3.10, which is why I have the code block:
```python
if sys.version_info < (3, 10):
    from importlib_metadata import entry_points
else:
    from importlib.metadata import entry_points
```
But I never added `importlib-metadata` (the distribution package that contains the import package `importlib_metadata`) to requirements.  So why did it ever pass?  Probably because, once upon a time, _another_ package required `importlib-metadata` it ended up in the lockfile, which you removed [here](https://github.com/andgoldschmidt/derivative/commit/818a14280a33579932084412a8b68dcb6708b7b5).  Which goes to show a benefit of not using the lockfile in CI: errors like this one (me forgetting to add importlib-metadata to pyproject.toml) didn't surface until the lockfile was removed.

Ok, so that explains the `ModuleNotFoundError`.  What about the `TypeError` for unexpected keyword argument?

Subtly, the versions of `entry_points()` that are in `importlib.metadata` and `importlib_metadata` differ:

importlib.metadata (3.9):
```
entry_points()
    Return EntryPoint objects for all installed packages.
    
    :return: EntryPoint objects for all installed packages.
```

importlib_metadata (current pypi version, 7.1.0):
```
entry_points(**params) -> 'EntryPoints'
    Return EntryPoint objects for all installed packages.
    
    Pass selection parameters (group or name) to filter the
    result to entry points matching those properties (see
    EntryPoints.select()).
    
    :return: EntryPoints for all installed packages.
```

So it appears that it's loading `entry_points` from `importlib.metadata`, even though it's python 3.9 and should've imported from `importlib_metadata`.  I'm not sure why that's happening, and I'll look into it.  As for not registering _any_ plugins, that might be an issue with your installation of `derivative`.  Try recreating the environment and reinstalling.  A fresh install on 3.9 gives me:
```python
>>> from importlib_metadata import entry_points
>>> entry_points(group="derivative.hyperparam_opt")
EntryPoints((EntryPoint(name='kalman.default', value='derivative.utils:_default_kalman', group='derivative.hyperparam_opt'),))
>>> 
```

**Comment by Jacob-Stevens-Haas** [comment]
Oh, I see.  The same commit that added `importlib-metadata` to pyproject.toml removed the guard that controlled where `entry_points()` was imported from, resulting in the 3.9 importlib.metadata rather than the backported importlib_metadata.  Reverting that (see #44) should fix the tests.

File: tests/test_interface.py
Comment (Jacob-Stevens-Haas) [inline tests/test_interface.py:109]: Hey Yash, can you make sure that this actually differentiates the correct axis, not just that it returns the correct shape?  Similar to how we did it in the pysindy test, or how it's done [in this test in this repo](https://github.com/andgoldschmidt/derivative/blob/2bfeb89b41b744e98565f4fb12aa8f6c028b505e/tests/test_interface.py#L37C5-L40C1)

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Thanks Yash for doing this!  One comment on the test, then just add `importlib-metadata` to the pyproject.toml (see andy's test commit).

File: pyproject.toml
Comment (Jacob-Stevens-Haas) [inline pyproject.toml:23]: no comment, no extra line

File: tests/test_interface.py
Comment (Jacob-Stevens-Haas) [inline tests/test_interface.py:112]: End on newline

**Review by Jacob-Stevens-Haas — COMMENTED** [review]
Two quick ones.  With #44, docs should pass

**Review by Jacob-Stevens-Haas — APPROVED** [review]

**Comment by Jacob-Stevens-Haas** [comment]
@andgoldschmidt Do you want a final review, or should we merge?

**Pull request merged**
